README.md
# MongoDB input plugin for Embulk

[![Build Status](https://travis-ci.org/hakobera/embulk-input-mongodb.svg)](https://travis-ci.org/hakobera/embulk-input-mongodb)

MongoDB input plugin for Embulk loads records from MongoDB.
This plugin loads documents as single-column records (column name is "record"). You can use filter plugins such as [embulk-filter-expand_json](https://github.com/civitaspo/embulk-filter-expand_json) or [embulk-filter-add_time](https://github.com/treasure-data/embulk-filter-add_time) to convert the json column to typed columns. [Rename filter](http://www.embulk.org/docs/built-in.html#rename-filter-plugin) is also useful to rename the typed columns.

## Overview

This plugin only works with embulk >= 0.8.8.

* **Plugin type**: input
* **Guess supported**: no

## Configuration

- Connection parameters
  One of them is required.
  
  - use MongoDB connection string URI
    - **uri**: [MongoDB connection string URI](http://docs.mongodb.org/manual/reference/connection-string/) (e.g. 'mongodb://localhost:27017/mydb') (string, required)
  - use separated URI parameters
    - **hosts**: list of hosts. `hosts` are pairs of host(string, required) and port(integer, optional, default: 27017)
    - **user**: (string, optional)
    - **password**:  (string, optional)
    - **database**:  (string, required)
- **collection**: source collection name (string, required)
- **fields**: **(deprecated)** ~~hash records that has the following two fields (array, required)~~
  ~~- name: Name of the column~~
  ~~- type: Column types as follows~~
    ~~- boolean~~
    ~~- long~~
    ~~- double~~
    ~~- string~~
    ~~- timestamp~~
- **id_field_name** Name of Object ID field name. Set if you want to change the default name `_id` (string, optional, default: "_id")
- **query**: A JSON document used for [querying](https://docs.mongodb.com/manual/tutorial/query-documents/) on the source collection. Documents are loaded from the colleciton if they match with this condition. (string, optional)
- **projection**: A JSON document used for [projection](https://docs.mongodb.com/manual/reference/operator/projection/positional/) on query results. Fields in a document are used only if they match with this condition. (string, optional)
- **sort**: Ordering of results (string, optional)
- **batch_size**: Limits the number of objects returned in one [batch](http://api.mongodb.com/java/current/com/mongodb/DBCursor.html#batchSize-int-) (integer, optional, default: 10000)
- **incremental_field** List of field name (list, optional, can't use with sort option)
- **last_record** Last loaded record for incremental load (hash, optional)
- **stop_on_invalid_record** Stop bulk load transaction if a document includes invalid record (such as unsupported object type) (boolean, optional, default: false)
- **json_column_name**: column name used in outputs (string, optional, default: "json")

## Example

### Exporting all objects

#### Specify with MongoDB connection string URI.

```yaml
in:
  type: mongodb
  uri: mongodb://myuser:mypassword@localhost:27017/my_database
  collection: "my_collection"
```

#### Specify with separated URI parameters.

```yaml
in:
  type: mongodb
  hosts:
  - {host: localhost, port: 27017}
  - {host: example.com, port: 27017}
  user: myuser
  password: mypassword
  database: my_database
  collection: "my_collection"
```

### Filtering documents by query and projection

```yaml
in:
  type: mongodb
  uri: mongodb://myuser:mypassword@localhost:27017/my_database
  collection: "my_collection"
  query: '{ field1: { $gte: 3 } }'
  projection: '{ "_id": 1, "field1": 1, "field2": 0 }'
  sort: '{ "field1": 1 }'
```

### Incremental loading

```yaml
in:
  type: mongodb
  uri: mongodb://myuser:mypassword@localhost:27017/my_database
  collection: "my_collection"
  query: '{ field1: { $gt: 3 } }'
  projection: '{ "_id": 1, "field1": 1, "field2": 1 }'
  incremental_field:
    - "field2"
  last_record: {"field2": 13215}
```

Plugin will create new query and sort value.
You can't use `incremental_field` option with `sort` option at the same time.

```
query { field1: { $gt: 3 }, field2: { $gt: 13215}}
sort {"field2", 1} # field2 ascending
```

You have to specify last_record with special characters when field type is `ObjectId` or `DateTime`.

```yaml
# ObjectId field
in:
  type: mongodb
  incremental_field:
    - "_id"
  last_record: {"_id": {"$oid": "5739b2261c21e58edfe39716"}}

# DateTime field
in:
  type: mongodb
  incremental_field:
    - "time_field"
  last_record: {"time_field": {"$date": "2015-01-25T13:23:15.000Z"}}
```

#### Run Incremental load

```
$ embulk run /path/to/config.yml -c config-diff.yml
```

### Advanced usage with filter plugins

```yaml
in:
  type: mongodb
  uri: mongodb://myuser:mypassword@localhost:27017/my_database
  collection: "my_collection"
  query: '{ "age": { $gte: 3 } }'
  projection: '{ "_id": 1, "age": 1, "ts": 1, "firstName": 1, "lastName": 1 }'

filters:
  # convert json column into typed columns
  - type: expand_json
    json_column_name: record
    expanded_columns:
      - {name: _id, type: long}
      - {name: ts, type: string}
      - {name: firstName, type: string}
      - {name: lastName, type: string}

  # rename column names
  - type: rename
    columns:
      _id: id
      firstName: first_name
      lastName: last_name

  # convert string "ts" column into timestamp "time" column
  - type: add_time
    from_column:
      name: ts
      timestamp_format: "%Y-%m-%dT%H:%M:%S.%N%z"
    to_column:
      name: time
      type: timestamp
```

## Build

```
$ ./gradlew gem
```

## Test

```
$ ./gradlew test  # -t to watch change of files and rebuild continuously
```

To run unit tests, we need to configure the following environment variables.

When environment variables are not set, skip almost test cases.

```
MONGO_URI
MONGO_COLLECTION
```

If you're using Mac OS X El Capitan and GUI Applications(IDE), like as follows.
```xml
$ vi ~/Library/LaunchAgents/environment.plist
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
<dict>
  <key>Label</key>
  <string>my.startup</string>
  <key>ProgramArguments</key>
  <array>
    <string>sh</string>
    <string>-c</string>
    <string>
      launchctl setenv MONGO_URI mongodb://myuser:mypassword@localhost:27017/my_database
      launchctl setenv MONGO_COLLECTION my_collection
    </string>
  </array>
  <key>RunAtLoad</key>
  <true/>
</dict>
</plist>

$ launchctl load ~/Library/LaunchAgents/environment.plist
$ launchctl getenv MONGO_URI //try to get value.

Then start your applications.
```


CHANGELOG.md
# Change Log

## [v0.5.0](https://github.com/hakobera/embulk-input-mongodb/tree/v0.5.0) (2016-08-23)
[Full Changelog](https://github.com/hakobera/embulk-input-mongodb/compare/v0.4.0...v0.5.0)

**Implemented enhancements:**

- Support separated URI params [\#24](https://github.com/hakobera/embulk-input-mongodb/pull/24) ([sakama](https://github.com/sakama))

**Merged pull requests:**

- Add --host option to mongoimport to prevent connection failure. [\#27](https://github.com/hakobera/embulk-input-mongodb/pull/27) ([hakobera](https://github.com/hakobera))
- Add document about batch\_size config [\#25](https://github.com/hakobera/embulk-input-mongodb/pull/25) ([hakobera](https://github.com/hakobera))

## [v0.4.0](https://github.com/hakobera/embulk-input-mongodb/tree/v0.4.0) (2016-07-08)
[Full Changelog](https://github.com/hakobera/embulk-input-mongodb/compare/v0.3.2...v0.4.0)

**Implemented enhancements:**

- Support incremental load [\#16](https://github.com/hakobera/embulk-input-mongodb/pull/16) ([sakama](https://github.com/sakama))

**Merged pull requests:**

- Bumpup version to 0.4.0 [\#23](https://github.com/hakobera/embulk-input-mongodb/pull/23) ([hakobera](https://github.com/hakobera))

## [v0.3.2](https://github.com/hakobera/embulk-input-mongodb/tree/v0.3.2) (2016-06-23)
[Full Changelog](https://github.com/hakobera/embulk-input-mongodb/compare/v0.3.1...v0.3.2)

**Implemented enhancements:**

- Support Symbol type [\#19](https://github.com/hakobera/embulk-input-mongodb/issues/19)

**Merged pull requests:**

- Bumpup version to 0.3.2 [\#22](https://github.com/hakobera/embulk-input-mongodb/pull/22) ([hakobera](https://github.com/hakobera))
- Release v0.3.1 [\#17](https://github.com/hakobera/embulk-input-mongodb/pull/17) ([hakobera](https://github.com/hakobera))
- Support Symbol type [\#21](https://github.com/hakobera/embulk-input-mongodb/pull/21) ([hakobera](https://github.com/hakobera))

## [v0.3.1](https://github.com/hakobera/embulk-input-mongodb/tree/v0.3.1) (2016-06-21)
[Full Changelog](https://github.com/hakobera/embulk-input-mongodb/compare/v0.3.0...v0.3.1)

**Implemented enhancements:**

- Support id\_field\_name config [\#15](https://github.com/hakobera/embulk-input-mongodb/pull/15) ([hakobera](https://github.com/hakobera))

**Closed issues:**

- readName can only be called when State is NAME, not when State is VALUE. [\#18](https://github.com/hakobera/embulk-input-mongodb/issues/18)

## [v0.3.0](https://github.com/hakobera/embulk-input-mongodb/tree/v0.3.0) (2016-06-13)
[Full Changelog](https://github.com/hakobera/embulk-input-mongodb/compare/v0.2.0...v0.3.0)

**Merged pull requests:**

- Add checkstyle plugin and format code [\#14](https://github.com/hakobera/embulk-input-mongodb/pull/14) ([sakama](https://github.com/sakama))
- Bumpup version to 0.3.0 [\#13](https://github.com/hakobera/embulk-input-mongodb/pull/13) ([hakobera](https://github.com/hakobera))
- Add unit test for existing code, \#9 and \#10 [\#11](https://github.com/hakobera/embulk-input-mongodb/pull/11) ([sakama](https://github.com/sakama))
- Change schema convert logic [\#10](https://github.com/hakobera/embulk-input-mongodb/pull/10) ([sakama](https://github.com/sakama))

## [v0.2.0](https://github.com/hakobera/embulk-input-mongodb/tree/v0.2.0) (2016-05-30)
[Full Changelog](https://github.com/hakobera/embulk-input-mongodb/compare/v0.1.2...v0.2.0)

**Implemented enhancements:**

- Support JSON type [\#6](https://github.com/hakobera/embulk-input-mongodb/issues/6)

**Merged pull requests:**

- Prepare release for v0.2.0 [\#12](https://github.com/hakobera/embulk-input-mongodb/pull/12) ([hakobera](https://github.com/hakobera))
- Add minimum limit to batch\_size and changed not to allow null [\#9](https://github.com/hakobera/embulk-input-mongodb/pull/9) ([sakama](https://github.com/sakama))
- Refactor some codes [\#8](https://github.com/hakobera/embulk-input-mongodb/pull/8) ([sakama](https://github.com/sakama))
- Fix example config on readme [\#5](https://github.com/hakobera/embulk-input-mongodb/pull/5) ([hakobera](https://github.com/hakobera))
- Support JSON type [\#7](https://github.com/hakobera/embulk-input-mongodb/pull/7) ([hakobera](https://github.com/hakobera))

## [v0.1.2](https://github.com/hakobera/embulk-input-mongodb/tree/v0.1.2) (2015-09-10)
[Full Changelog](https://github.com/hakobera/embulk-input-mongodb/compare/v0.1.1...v0.1.2)

**Merged pull requests:**

- Change build option for JDK7 [\#4](https://github.com/hakobera/embulk-input-mongodb/pull/4) ([hakobera](https://github.com/hakobera))
- Set homepage info [\#3](https://github.com/hakobera/embulk-input-mongodb/pull/3) ([hakobera](https://github.com/hakobera))

## [v0.1.1](https://github.com/hakobera/embulk-input-mongodb/tree/v0.1.1) (2015-09-06)
[Full Changelog](https://github.com/hakobera/embulk-input-mongodb/compare/v0.1.0...v0.1.1)

## [v0.1.0](https://github.com/hakobera/embulk-input-mongodb/tree/v0.1.0) (2015-09-06)
**Implemented enhancements:**

- Fix projection logic and better logging [\#2](https://github.com/hakobera/embulk-input-mongodb/pull/2) ([hakobera](https://github.com/hakobera))

**Merged pull requests:**

- Run test on travis [\#1](https://github.com/hakobera/embulk-input-mongodb/pull/1) ([hakobera](https://github.com/hakobera))



\* *This Change Log was automatically generated by [github_changelog_generator](https://github.com/skywinder/Github-Changelog-Generator)*

src/main/java/org/embulk/input/mongodb/ValueCodec.java
package org.embulk.input.mongodb;

import org.bson.BsonReader;
import org.bson.BsonType;
import org.bson.BsonWriter;
import org.bson.codecs.Codec;
import org.bson.codecs.DecoderContext;
import org.bson.codecs.EncoderContext;
import org.embulk.spi.DataException;
import org.embulk.spi.Exec;
import org.msgpack.value.Value;
import org.slf4j.Logger;

import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Date;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.TimeZone;

import static org.msgpack.value.ValueFactory.*;

public class ValueCodec implements Codec<Value>
{
    private final SimpleDateFormat formatter;
    private final Logger log = Exec.getLogger(MongodbInputPlugin.class);
    private final boolean stopOnInvalidRecord;
    private final MongodbInputPlugin.PluginTask task;

    public ValueCodec(boolean stopOnInvalidRecord, MongodbInputPlugin.PluginTask task)
    {
        this.formatter = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'", java.util.Locale.ENGLISH);
        formatter.setTimeZone(TimeZone.getTimeZone("UTC"));
        this.stopOnInvalidRecord = stopOnInvalidRecord;
        this.task = task;
    }

    @Override
    public void encode(final BsonWriter writer, final Value value, final EncoderContext encoderContext)
    {
        throw new UnsupportedOperationException();
    }

    @Override
    public Value decode(final BsonReader reader, final DecoderContext decoderContext)
    {
        Map<Value, Value> kvs = new LinkedHashMap<>();

        reader.readStartDocument();
        while (reader.readBsonType() != BsonType.END_OF_DOCUMENT) {
            String fieldName = reader.readName();
            BsonType type = reader.getCurrentBsonType();
            fieldName = normalize(fieldName);

            try {
                kvs.put(newString(fieldName), readValue(reader, decoderContext));
            }
            catch (UnknownTypeFoundException ex) {
                reader.skipValue();
                if (stopOnInvalidRecord) {
                    throw ex;
                }
                log.warn(String.format("Skipped document because field '%s' contains unsupported object type [%s]",
                        fieldName, type));
            }
        }
        reader.readEndDocument();

        return newMap(kvs);
    }

    public Value decodeArray(final BsonReader reader, final DecoderContext decoderContext)
    {
        List<Value> list = new ArrayList<>();

        reader.readStartArray();
        while (reader.readBsonType() != BsonType.END_OF_DOCUMENT) {
            list.add(readValue(reader, decoderContext));
        }
        reader.readEndArray();

        return newArray(list);
    }

    private Value readValue(BsonReader reader, DecoderContext decoderContext)
    {
        switch (reader.getCurrentBsonType()) {
            // https://docs.mongodb.com/manual/reference/bson-types/
            // https://github.com/mongodb/mongo-java-driver/tree/master/bson/src/main/org/bson/codecs
            case DOUBLE:
                return newFloat(reader.readDouble());
            case STRING:
                return newString(reader.readString());
            case ARRAY:
                return decodeArray(reader, decoderContext);
            case BINARY:
                return newBinary(reader.readBinaryData().getData(), true);
            case OBJECT_ID:
                return newString(reader.readObjectId().toString());
            case BOOLEAN:
                return newBoolean(reader.readBoolean());
            case DATE_TIME:
                return newString(formatter.format(new Date(reader.readDateTime())));
            case NULL:
                reader.readNull();
                return newNil();
            case REGULAR_EXPRESSION:
                return newString(reader.readRegularExpression().toString());
            case JAVASCRIPT:
                return newString(reader.readJavaScript());
            case JAVASCRIPT_WITH_SCOPE:
                return newString(reader.readJavaScriptWithScope());
            case INT32:
                return newInteger(reader.readInt32());
            case TIMESTAMP:
                return newInteger(reader.readTimestamp().getTime());
            case INT64:
                return newInteger(reader.readInt64());
            case DOCUMENT:
                return decode(reader, decoderContext);
            case SYMBOL:
                return newString(reader.readSymbol());
            default: // e.g. MIN_KEY, MAX_KEY, DB_POINTER, UNDEFINED
                throw new UnknownTypeFoundException(String.format("Unsupported type %s of '%s' field. Please exclude the field from 'projection:' option",
                        reader.getCurrentBsonType(), reader.getCurrentName()));
        }
    }

    @Override
    public Class<Value> getEncoderClass()
    {
        return Value.class;
    }

    private String normalize(String key)
    {
        if (key.equals("_id")) {
            return task.getIdFieldName();
        }
        return key;
    }

    public static class UnknownTypeFoundException extends DataException
    {
        UnknownTypeFoundException(String message)
        {
            super(message);
        }
    }
}


src/main/java/org/embulk/input/mongodb/MongodbInputPlugin.java
package org.embulk.input.mongodb;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.google.common.base.Optional;
import com.google.common.base.Throwables;
import com.mongodb.MongoClient;
import com.mongodb.MongoClientURI;
import com.mongodb.MongoCredential;
import com.mongodb.MongoException;
import com.mongodb.ServerAddress;
import com.mongodb.client.MongoCollection;
import com.mongodb.client.MongoCursor;
import com.mongodb.client.MongoDatabase;
import com.mongodb.util.JSON;
import com.mongodb.util.JSONParseException;
import org.bson.codecs.configuration.CodecRegistries;
import org.bson.codecs.configuration.CodecRegistry;
import org.bson.conversions.Bson;
import org.embulk.config.Config;
import org.embulk.config.ConfigDefault;
import org.embulk.config.ConfigDiff;
import org.embulk.config.ConfigException;
import org.embulk.config.ConfigInject;
import org.embulk.config.ConfigSource;
import org.embulk.config.DataSource;
import org.embulk.config.DataSourceImpl;
import org.embulk.config.ModelManager;
import org.embulk.config.Task;
import org.embulk.config.TaskReport;
import org.embulk.config.TaskSource;
import org.embulk.spi.BufferAllocator;
import org.embulk.spi.Column;
import org.embulk.spi.Exec;
import org.embulk.spi.InputPlugin;
import org.embulk.spi.PageBuilder;
import org.embulk.spi.PageOutput;
import org.embulk.spi.Schema;
import org.embulk.spi.SchemaConfig;
import org.embulk.spi.type.Types;
import org.msgpack.value.Value;
import org.slf4j.Logger;

import javax.validation.constraints.Min;

import java.io.IOException;
import java.net.UnknownHostException;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

public class MongodbInputPlugin
        implements InputPlugin
{
    public interface HostTask
            extends Task
    {
        @Config("host")
        String getHost();

        @Config("port")
        @ConfigDefault("27017")
        int getPort();
    }

    public interface PluginTask
            extends Task
    {
        // MongoDB connection string URI
        @Config("uri")
        @ConfigDefault("null")
        Optional<String> getUri();

        @Config("hosts")
        @ConfigDefault("null")
        Optional<List<HostTask>> getHosts();

        @Config("user")
        @ConfigDefault("null")
        Optional<String> getUser();

        @Config("password")
        @ConfigDefault("null")
        Optional<String> getPassword();

        @Config("database")
        @ConfigDefault("null")
        Optional<String> getDatabase();

        @Config("collection")
        String getCollection();

        @Config("fields")
        @ConfigDefault("null")
        Optional<SchemaConfig> getFields();

        @Config("projection")
        @ConfigDefault("\"{}\"")
        String getProjection();

        @Config("query")
        @ConfigDefault("\"{}\"")
        String getQuery();
        void setQuery(String query);

        @Config("sort")
        @ConfigDefault("\"{}\"")
        String getSort();
        void setSort(String sort);

        @Config("id_field_name")
        @ConfigDefault("\"_id\"")
        String getIdFieldName();

        @Config("batch_size")
        @ConfigDefault("10000")
        @Min(1)
        int getBatchSize();

        @Config("stop_on_invalid_record")
        @ConfigDefault("false")
        boolean getStopOnInvalidRecord();

        @Config("json_column_name")
        @ConfigDefault("\"record\"")
        String getJsonColumnName();

        @Config("incremental_field")
        @ConfigDefault("null")
        Optional<List<String>> getIncrementalField();

        @Config("last_record")
        @ConfigDefault("null")
        Optional<Map<String, Object>> getLastRecord();

        @ConfigInject
        BufferAllocator getBufferAllocator();
    }

    private final Logger log = Exec.getLogger(MongodbInputPlugin.class);

    @Override
    public ConfigDiff transaction(ConfigSource config,
            InputPlugin.Control control)
    {
        PluginTask task = config.loadConfig(PluginTask.class);
        if (task.getFields().isPresent()) {
            throw new ConfigException("field option was deprecated so setting will be ignored");
        }
        if (task.getIncrementalField().isPresent() && !task.getSort().equals("{}")) {
            throw new ConfigException("both of sort and incremental_load can't be used together");
        }

        Map<String, String> newCondition = buildIncrementalCondition(task);
        task.setQuery(newCondition.get("query"));
        task.setSort(newCondition.get("sort"));

        validateJsonField("projection", task.getProjection());
        validateJsonField("query", task.getQuery());
        validateJsonField("sort", task.getSort());

        // Connect once to throw ConfigException in earlier stage of excecution
        try {
            connect(task);
        }
        catch (UnknownHostException | MongoException ex) {
            throw new ConfigException(ex);
        }
        Schema schema = Schema.builder().add(task.getJsonColumnName(), Types.JSON).build();
        return resume(task.dump(), schema, 1, control);
    }

    @Override
    public ConfigDiff resume(TaskSource taskSource,
            Schema schema, int taskCount,
            InputPlugin.Control control)
    {
        List<TaskReport> report = control.run(taskSource, schema, taskCount);

        ConfigDiff configDiff = Exec.newConfigDiff();
        if (report.size() > 0 && report.get(0).has("last_record")) {
            configDiff.set("last_record", report.get(0).get(Map.class, "last_record"));
        }

        return configDiff;
    }

    @Override
    public void cleanup(TaskSource taskSource,
            Schema schema, int taskCount,
            List<TaskReport> successCommitReports)
    {
        // do nothing
    }

    @Override
    public TaskReport run(TaskSource taskSource,
            Schema schema, int taskIndex,
            PageOutput output)
    {
        PluginTask task = taskSource.loadTask(PluginTask.class);
        BufferAllocator allocator = task.getBufferAllocator();
        PageBuilder pageBuilder = new PageBuilder(allocator, schema, output);
        final Column column = pageBuilder.getSchema().getColumns().get(0);

        ValueCodec valueCodec = new ValueCodec(task.getStopOnInvalidRecord(), task);
        MongoCollection<Value> collection;
        try {
            MongoDatabase db = connect(task);

            CodecRegistry registry = CodecRegistries.fromRegistries(
                    MongoClient.getDefaultCodecRegistry(),
                    CodecRegistries.fromCodecs(valueCodec)
            );
            collection = db.getCollection(task.getCollection(), Value.class)
                    .withCodecRegistry(registry);
        }
        catch (UnknownHostException | MongoException ex) {
            throw new ConfigException(ex);
        }

        Bson query = (Bson) JSON.parse(task.getQuery());
        Bson projection = (Bson) JSON.parse(task.getProjection());
        Bson sort = (Bson) JSON.parse(task.getSort());

        log.trace("query: {}", query);
        log.trace("projection: {}", projection);
        log.trace("sort: {}", sort);

        try (MongoCursor<Value> cursor = collection
                .find(query)
                .projection(projection)
                .sort(sort)
                .batchSize(task.getBatchSize())
                .iterator()) {
            while (cursor.hasNext()) {
                pageBuilder.setJson(column, cursor.next());
                pageBuilder.addRecord();
            }
        } catch (MongoException ex) {
            Throwables.propagate(ex);
        }

        pageBuilder.finish();

        TaskReport report = Exec.newTaskReport();

        if (valueCodec.getLastRecord() != null) {
            DataSource lastRecord = new DataSourceImpl(Exec.getInjector().getInstance(ModelManager.class));
            for (String k : valueCodec.getLastRecord().keySet()) {
                String value = valueCodec.getLastRecord().get(k).toString();
                Map<String, String> types = valueCodec.getLastRecordType();
                HashMap<String, String> innerValue = new HashMap<>();
                switch(types.get(k)) {
                    case "OBJECT_ID":
                        innerValue.put("$oid", value);
                        lastRecord.set(k, innerValue);
                        break;
                    case "DATE_TIME":
                        innerValue.put("$date", value);
                        lastRecord.set(k, innerValue);
                        break;
                    case "INT32":
                    case "INT64":
                    case "TIMESTAMP":
                        lastRecord.set(k, Integer.valueOf(value));
                        break;
                    case "BOOLEAN":
                        lastRecord.set(k, Boolean.valueOf(value));
                        break;
                    case "DOUBLE":
                        lastRecord.set(k, Double.valueOf(value));
                        break;
                    case "DOCUMENT":
                    case "ARRAY":
                        throw new ConfigException(String.format("Unsupported type '%s' was given for 'last_record' [%s]", types.get(k), value));
                    default:
                        lastRecord.set(k, value);
                }
            }
            report.setNested("last_record", lastRecord);
        }
        return report;
    }

    @Override
    public ConfigDiff guess(ConfigSource config)
    {
        return Exec.newConfigDiff();
    }

    private MongoDatabase connect(final PluginTask task) throws UnknownHostException, MongoException
    {
        MongoClient mongoClient;
        String database;

        if (!task.getUri().isPresent() && !task.getHosts().isPresent()) {
            throw new ConfigException("'uri' or 'hosts' is required");
        }

        if (task.getUri().isPresent()) {
            MongoClientURI uri = new MongoClientURI(task.getUri().get());
            database = uri.getDatabase();
            mongoClient = new MongoClient(uri);
        }
        else {
            mongoClient = createClientFromParams(task);
            database = task.getDatabase().get();
        }

        MongoDatabase db = mongoClient.getDatabase(database);
        // Get collection count for throw Exception
        db.getCollection(task.getCollection()).count();
        return db;
    }

    private MongoClient createClientFromParams(PluginTask task)
    {
        if (!task.getHosts().isPresent()) {
            throw new ConfigException("'hosts' option's value is required but empty");
        }
        if (!task.getDatabase().isPresent()) {
            throw new ConfigException("'database' option's value is required but empty");
        }

        List<ServerAddress> addresses = new ArrayList<>();
        for (HostTask host : task.getHosts().get()) {
            addresses.add(new ServerAddress(host.getHost(), host.getPort()));
        }

        if (task.getUser().isPresent()) {
            MongoCredential credential = MongoCredential.createCredential(
                    task.getUser().get(),
                    task.getDatabase().get(),
                    task.getPassword().get().toCharArray()
            );
            return new MongoClient(addresses, Arrays.asList(credential));
        }
        else {
            return new MongoClient(addresses);
        }
    }

    private Map<String, String> buildIncrementalCondition(PluginTask task)
    {
        Map<String, String> result = new HashMap<>();
        String query = task.getQuery();
        String sort = task.getSort();
        result.put("query", query);
        result.put("sort", sort);

        Optional<List<String>> incrementalField = task.getIncrementalField();
        Optional<Map<String, Object>> lastRecord = task.getLastRecord();
        if (!incrementalField.isPresent()) {
            return result;
        }

        Map<String, Object> newQuery = new LinkedHashMap<>();
        Map<String, Integer> newSort = new LinkedHashMap<>();
        ObjectMapper mapper = new ObjectMapper();

        try {
            @SuppressWarnings("unchecked")
            Map<String, Object> queryJson = mapper.readValue(query, Map.class);
            for (String k : queryJson.keySet()) {
                newQuery.put(k, queryJson.get(k));
            }

            if (lastRecord.isPresent()) {
                for (String k : lastRecord.get().keySet()) {
                    Map<String, Object> v = new HashMap<>();
                    Object record = lastRecord.get().get(k);

                    if (newQuery.containsKey(k)) {
                        throw new ConfigException("Field declaration was duplicated between 'incremental_field' and 'query' options");
                    }

                    v.put("$gt", record);
                    newQuery.put(k, v);
                }
                String newQueryString = mapper.writeValueAsString(newQuery);
                log.info(String.format("New query value was generated for incremental load: '%s'", newQueryString));
                result.put("query", newQueryString);
            }

            for (String k : incrementalField.get()) {
                newSort.put(k, 1);
            }

            String newSortString = mapper.writeValueAsString(newSort);
            log.info(String.format("New sort value was generated for incremental load: '%s'", newSortString));
            result.put("sort", newSortString);

            return result;
        }
        catch (JSONParseException | IOException ex) {
            throw new ConfigException("Could not generate new query for incremental load.");
        }
    }

    private void validateJsonField(String name, String jsonString)
    {
        try {
            JSON.parse(jsonString);
        }
        catch (JSONParseException ex) {
            throw new ConfigException(String.format("Invalid JSON string was given for '%s' parameter. [%s]", name, jsonString));
        }
    }
}
